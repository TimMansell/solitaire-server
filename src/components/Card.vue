<template>
  <div
    class="card"
    :class="{'card--is-visible': visible}">
    <span v-if="visible">{{ value }}{{ suit }}</span>
  </div>
</template>

<script>
export default {
  name: 'Card',
  props: {
    value: {
      type: String,
      default: 'A',
    },
    suit: {
      type: String,
      default: 'c',
    },
    visible: {
      type: Boolean,
      default: false,
    },
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped lang="scss">
.card {
  display: flex;
  flex: 1;
  justify-content: center;
  border: 1px solid black;
  background: grey;
  width: 100%;
  height: 100px;
  border-radius: 5px;

  &:nth-of-type(n+2) {
    margin-top: -50%;
  }

  &--is-visible {
    background: white;
  }
}
</style>
