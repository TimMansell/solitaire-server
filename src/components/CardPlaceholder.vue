<template>
  <div :class="classes">
    <SvgIcon data-test="card-placeholder" name="Card_back_17" />
  </div>
</template>

<script>
import SvgIcon from '@/components/SvgIcon.vue';

export default {
  name: 'CardPlaceholder',
  components: {
    SvgIcon,
  },
  props: {
    seeThrough: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    classes() {
      const { seeThrough } = this;

      return {
        'card-placeholder': seeThrough,
      };
    },
  },
};
</script>

<style scoped lang="scss">
.card-placeholder {
  opacity: 0.15;
}
</style>
