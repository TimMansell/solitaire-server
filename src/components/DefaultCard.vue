<template>
  <SvgIcon :name="cardName" :width="width" />
</template>

<script>
import SvgIcon from '@/components/SvgIcon.vue';

export default {
  name: 'Card',
  components: {
    SvgIcon,
  },
  props: {
    value: {
      type: String,
      default: 'A',
    },
    suit: {
      type: String,
      default: 'c',
    },
    width: {
      type: Number,
      default: 0,
    },
  },
  computed: {
    cardName() {
      const { suit, value } = this;

      return `${value}${suit.toUpperCase()}`;
    },
  },
};
</script>
