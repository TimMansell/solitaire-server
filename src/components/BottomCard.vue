<template>
  <TouchEvents
    @swipe="autoMoveCard($event, id)"
    @doubletap="autoMoveCard($event, id)"
  >
    <DefaultCard :value="value" />
  </TouchEvents>
</template>

<script>
import { mapActions } from 'vuex';
import DefaultCard from '@/components/DefaultCard.vue';
import TouchEvents from '@/components/TouchEvents.vue';

export default {
  name: 'Card',
  components: {
    DefaultCard,
    TouchEvents,
  },
  props: {
    id: {
      type: [String, Number],
      default: 0,
    },
    value: {
      type: String,
      default: 'Aâ™£',
    },
  },
  methods: {
    ...mapActions(['autoMoveCardToFoundation']),
    autoMoveCard(e, id) {
      this.autoMoveCardToFoundation(id);
    },
  },
};
</script>
