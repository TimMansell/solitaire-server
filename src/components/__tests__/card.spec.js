import { shallowMount } from '@vue/test-utils';
import Card from '@/components/Card.vue';
import DefaultCard from '@/components/DefaultCard.vue';
import BottomCard from '@/components/BottomCard.vue';
import CardPlaceholder from '@/components/CardPlaceholder.vue';

const computed = {
  selectedCardId: () => null,
};

describe('Card.vue', () => {
  it('matches visible snapshot', () => {
    const wrapper = shallowMount(Card, {
      computed,
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('matches hidden snapshot', () => {
    const wrapper = shallowMount(Card, {
      propsData: {
        visible: false,
      },
      computed,
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('matches bottom card snapshot', () => {
    const wrapper = shallowMount(Card, {
      propsData: {
        bottomCard: true,
      },
      computed,
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('should have correct props', () => {
    const propsData = {
      id: 2,
      value: 'K',
      suit: '♦',
      visible: true,
      revealed: true,
      clickable: false,
      bottomCard: true,
      stacked: true,
    };

    const wrapper = shallowMount(Card, {
      propsData,
      computed,
    });

    expect(wrapper.props().id).toBe(2);
    expect(wrapper.props().value).toBe(propsData.value);
    expect(wrapper.props().suit).toBe(propsData.suit);
    expect(wrapper.props().visible).toBe(true);
    expect(wrapper.props().revealed).toBe(true);
    expect(wrapper.props().clickable).toBe(false);
    expect(wrapper.props().bottomCard).toBe(true);
    expect(wrapper.props().stacked).toBe(true);
  });

  it('should render a diamond card', () => {
    const propsData = {
      value: '6',
      suit: '♦',
    };

    const wrapper = shallowMount(Card, {
      propsData,
      computed,
    });

    expect(wrapper.findComponent(DefaultCard).exists()).toBe(true);
    expect(wrapper.attributes('data-test')).toBe(
      `card-${propsData.value}${propsData.suit}`
    );
  });

  it('should render a club card', () => {
    const propsData = {
      value: '6',
      suit: '♣',
    };

    const wrapper = shallowMount(Card, {
      propsData,
      computed,
    });

    expect(wrapper.findComponent(DefaultCard).exists()).toBe(true);
    expect(wrapper.attributes('data-test')).toBe(
      `card-${propsData.value}${propsData.suit}`
    );
  });

  it('should render a heart card', () => {
    const propsData = {
      value: '6',
      suit: '♥',
    };

    const wrapper = shallowMount(Card, {
      propsData,
      computed,
    });

    expect(wrapper.findComponent(DefaultCard).exists()).toBe(true);
    expect(wrapper.attributes('data-test')).toBe(
      `card-${propsData.value}${propsData.suit}`
    );
  });

  it('should render a spade card', () => {
    const propsData = {
      value: '6',
      suit: '♠',
    };

    const wrapper = shallowMount(Card, {
      propsData,
      computed,
    });

    expect(wrapper.findComponent(DefaultCard).exists()).toBe(true);
    expect(wrapper.attributes('data-test')).toBe(
      `card-${propsData.value}${propsData.suit}`
    );
  });

  it('should render a default card', () => {
    const wrapper = shallowMount(Card, {
      computed,
    });

    expect(wrapper.findComponent(DefaultCard).exists()).toBe(true);
  });

  it('should render a bottom card', () => {
    const wrapper = shallowMount(Card, {
      propsData: {
        bottomCard: true,
      },
      computed,
    });

    expect(wrapper.findComponent(BottomCard).exists()).toBe(true);
  });

  it('should render a hidden card', () => {
    const wrapper = shallowMount(Card, {
      propsData: {
        visible: false,
      },
      computed,
    });

    expect(wrapper.attributes('data-test')).toBe('card-hidden');
    expect(wrapper.findComponent(CardPlaceholder).exists()).toBe(true);
  });

  it('should render a stacked card', () => {
    const wrapper = shallowMount(Card, {
      propsData: {
        stacked: true,
      },
      computed,
    });

    expect(wrapper.classes()).toContain('card--is-stacked');
  });

  it('should render a selected card', () => {
    const wrapper = shallowMount(Card, {
      computed: {
        selectedCardId: () => 0,
      },
    });

    expect(wrapper.classes()).toContain('card--is-selected');
  });

  describe('Set Card', () => {
    const mockStore = { dispatch: jest.fn() };

    const mocks = {
      $store: mockStore,
    };

    it('should not call store action "setCard" when clicked', () => {
      const wrapper = shallowMount(Card, {
        mocks,
        propsData: {
          clickable: false,
        },
        computed,
      });

      wrapper.trigger('click');

      expect(wrapper.classes()).toContain('card--is-not-clickable');
      expect(mockStore.dispatch).toHaveBeenCalledTimes(0);
    });

    it('should not call store action "setCard" when clicked when not visible', () => {
      const wrapper = shallowMount(Card, {
        mocks,
        propsData: {
          visible: false,
        },
        computed,
      });

      wrapper.trigger('click');

      expect(mockStore.dispatch).toHaveBeenCalledTimes(0);
    });

    it('should call store action "setCard" when clicked when visible', () => {
      const wrapper = shallowMount(Card, {
        mocks,
        computed,
      });

      wrapper.trigger('click');

      expect(mockStore.dispatch).toHaveBeenCalledTimes(1);
    });
  });
});
