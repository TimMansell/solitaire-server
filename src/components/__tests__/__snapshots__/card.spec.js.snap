// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card.vue matches bottom card snapshot 1`] = `
<div draggable="true" data-test="card-K♦" class="card card--is-draggable">
  <defaultcard-stub value="K♦" style="display: none;"></defaultcard-stub>
  <bottomcard-stub id="2" value="K♦"></bottomcard-stub>
  <cardplaceholder-stub cards="1" style="display: none;"></cardplaceholder-stub>
</div>
`;

exports[`Card.vue matches hidden snapshot 1`] = `
<div draggable="false" data-test="card-hidden" class="card">
  <defaultcard-stub value="K♦" style="display: none;"></defaultcard-stub>
  <bottomcard-stub id="2" value="K♦" style="display: none;"></bottomcard-stub>
  <cardplaceholder-stub cards="1"></cardplaceholder-stub>
</div>
`;

exports[`Card.vue matches visible snapshot 1`] = `
<div draggable="true" data-test="card-K♦" class="card card--is-draggable">
  <defaultcard-stub value="K♦"></defaultcard-stub>
  <bottomcard-stub id="2" value="K♦" style="display: none;"></bottomcard-stub>
  <cardplaceholder-stub cards="1" style="display: none;"></cardplaceholder-stub>
</div>
`;
